source/common/http/header_map_impl.cc

```diff
@@ -649,5 +655,410 @@ HeaderMapImplUtility::getAllHeaderMapImplInfo() {
+  if (entry) {
+    const uint64_t added_size = appendToHeader(entry->value(), value);
+    addSize(added_size);
+  } else {
+    addCopy(key, value);
+  }
+}
+
+void HeaderMapImpl::setReference(const LowerCaseString& key, absl::string_view value) {
+  HeaderString ref_key(key);
+  HeaderString ref_value(value);
+  remove(key);
+  addViaMove(std::move(ref_key), std::move(ref_value));
+}
+
+void HeaderMapImpl::setReferenceKey(const LowerCaseString& key, absl::string_view value) {
+  HeaderString ref_key(key);
+  HeaderString new_value;
+  new_value.setCopy(value);
+  remove(key);
+  addViaMove(std::move(ref_key), std::move(new_value));
+  ASSERT(new_value.empty()); // NOLINT(bugprone-use-after-move)
+}
+
+void HeaderMapImpl::setCopy(const LowerCaseString& key, absl::string_view value) {
+  // Replaces the first occurrence of a header if it exists, otherwise adds by copy.
+  // TODO(#9221): converge on and document a policy for coalescing multiple headers.
+  auto* entry = getExisting(key);
+  if (entry) {
+    updateSize(entry->value().size(), value.size());
+    entry->value(value);
+  } else {
+    addCopy(key, value);
+  }
+}
+
+uint64_t HeaderMapImpl::byteSize() const { return cached_byte_size_; }
+
+void HeaderMapImpl::verifyByteSizeInternalForTest() const {
+  // Computes the total byte size by summing the byte size of the keys and values.
+  uint64_t byte_size = 0;
+  for (const auto& map_it : headers_map_) {
+    for (const HeaderEntryImpl& header : map_it.second) {
+      byte_size += header.key().size();
+      byte_size += header.value().size();
+    }
+  }
+  ASSERT(cached_byte_size_ == byte_size);
+}
+
+const HeaderEntry* HeaderMapImpl::get(const LowerCaseString& key) const {
+  // Returns the first occurrence of a Header (if any)
+  const HeaderEagerMapValue* value_list = findMapElement(key);
+  if (value_list == nullptr) {
+    return nullptr;
+  }
+
+  return &value_list->front();
+}
+
+void HeaderMapImpl::mapToSortedVector(const HeaderEagerMap& map,
+                                      std::vector<const HeaderEntryImpl*>& output_vector) const {
+  ASSERT(output_vector.empty());
+  // Add all HeaderEntryImpl elements, and then sort the vector
+  for (const auto& map_entry : map) {
+    for (const HeaderEntryImpl& header_entry : map_entry.second) {
+      output_vector.push_back(&header_entry);
+    }
+  }
+
+  std::sort(output_vector.begin(), output_vector.end(),
+            [](const HeaderEntryImpl* a, const HeaderEntryImpl* b) -> bool {
+              // The iteration order is as follows:
+              // 1. pseudo headers
+              // 2. non-pseudo headers
+              // Within each group, the original header addition order is preserved
+              const bool a_pseudo_header =
+                  !a->key().getStringView().empty() && a->key().getStringView()[0] == ':';
+              const bool b_pseudo_header =
+                  !b->key().getStringView().empty() && b->key().getStringView()[0] == ':';
```

Comment:
Can we simplify this by left-shifting all non-pseudo-headers by 8 or 16 bits, so this comparator can just be `return a->index_ < b->index_;` ?